src/python/pants/core/util_rules/environments.py

```diff
@@ -374,6 +375,66 @@ class RemoteEnvironmentTarget(Target):
 # -------------------------------------------------------------------------------------------
 
 
+@rule_helper
+async def _warn_on_non_local_environments(specified_targets: Iterable[Target], source: str) -> None:
```

Comment:
This would be good to test extensively with unit tests. you can do that by using `run_rule_with_mocks`. 

For error message helper functions, we actually intentionally do exact string matches to make sure we get everything right, including formatting. Re: that being brittle, if someone wants to change the error message, then they simply update the test.

Reply:
ack

Reply:
@Eric-Arellano Tests have been added; currently I'm testing for substrings because `parametrize` is more amenable to that. If you feel strongly about it, I can add a test that does a full string match.